File: C-SDL2/score.c
#include "score.h"

bool score_new(struct Score **score, SDL_Renderer *renderer) {

    *score = calloc(1, sizeof(struct Score));
    if (!*score) {
        fprintf(stderr, "Error in calloc of score!\n");
        return true;
    }
    struct Score *s = *score;

    s->renderer = renderer;
    s->color = (SDL_Color){255, 255, 255, 255};

    s->font = TTF_OpenFont("fonts/freesansbold.ttf", FONT_SIZE);
    if (!s->font) {
        fprintf(stderr, "Error creating font: %s\n", TTF_GetError());
        return true;
    }

    s->rect.x = 10;
    s->rect.y = 10;

    if (score_reset(s)) {
        return true;
    }

    return false;
}

void score_free(struct Score **score) {
    if (*score) {
        TTF_CloseFont((*score)->font);
        (*score)->font = NULL;
        SDL_DestroyTexture((*score)->image);
        (*score)->image = NULL;
        (*score)->renderer = NULL;
        free(*score);
        *score = NULL;
    }
}

bool score_reset(struct Score *s) {
    s->score = 0;
    if (score_update(s)) {
        return true;
    }

    return false;
}

bool score_update(struct Score *s) {
    if (s->image) {
        SDL_DestroyTexture(s->image);
    }
    
    int length = snprintf(NULL, 0, "Score: %d", s->score) + 1;
    char score_str[length];
    snprintf(score_str, (size_t)length, "Score: %d", s->score);

    SDL_Surface *surface = TTF_RenderText_Blended(s->font, score_str, s->color);
    if (!surface) {
        fprintf(stderr, "Error creating a surface: %s\n", SDL_GetError());
        return true;
    }

    s->image = SDL_CreateTextureFromSurface(s->renderer, surface);
    if (!s->image) {
        fprintf(stderr, "Error creating a texture: %s\n", SDL_GetError());
        return true;
    }

    SDL_FreeSurface(surface);

    if (SDL_QueryTexture(s->image, NULL, NULL, &s->rect.w, &s->rect.h)) {
        fprintf(stderr, "Error while querying texture: %s\n", SDL_GetError());
        return true;
    }
    
    return false;
}

bool score_increment(struct Score *s) {
    s->score++;
    if (score_update(s)) {
        return true;
    }

    return false;
}

bool score_draw(struct Score *s) {
    if (SDL_RenderCopy(s->renderer, s->image, NULL, &s->rect)) {
        fprintf(stderr, "Error while rendering texture: %s\n", SDL_GetError());
        return true;
    }

    return false;
}

File: Gforth-SDL2/score.fs
VARIABLE score-image
CREATE score-rect SDL_Rect ALLOT
CREATE score-color SDL_Color ALLOT
24 CONSTANT score-font-size
VARIABLE score-font
VARIABLE score 0 score !

: score-cleanup ( -- )
    score-font @ TTF_CloseFont
    score-image @ SDL_DestroyTexture
;

: score-update ( -- error )
    score-image @ SDL_DestroyTexture
    0 score-image !
    score-font @
    s" Score: " PAD PLACE
    score @ 0 <<# #s #> #>> PAD +PLACE
    s\" \0" PAD +PLACE
    PAD COUNT DROP
    score-color TTF_RenderText_Blended DUP @ 0= IF
        ." Error creating font surface: " SDL_GetError c-str> TYPE CR
        DROP
        TRUE
    ELSE
        score-rect rect-from-surface
        10 score-rect SDL_Rect-x int32>!
        10 score-rect SDL_Rect-y int32>!
        score-image SWAP texture-from-surface IF
            TRUE
        ELSE
            FALSE
        THEN
    THEN
;

: score-reset ( -- error )
    0 score !
    score-update
;

: score-increment ( -- error )
    1 score @ + score !
    score-update
;

: score-init ( -- error )
    score-color
    255 OVER SDL_Color-r C!
    255 OVER SDL_Color-g C!
    255 OVER SDL_Color-b C!
    255 SWAP SDL_Color-a C!

    s" fonts/freesansbold.ttf" >c-str score-font-size TTF_OpenFont score-font !
    score-font @ 0= IF
        ." Error creating font: " SDL_GetError c-str> TYPE CR
        FALSE
    ELSE
        score-update IF
            FALSE
        ELSE
            TRUE
        THEN
    THEN
;

: score-draw ( -- )
    renderer @ score-image @ 0 score-rect SDL_RenderCopy DROP
;

